### PR TITLE
Gh 48 show link text

### PR DESCRIPTION
**Pull request by dantleech**
for example:

```
{
  "url": "https://www.bbc.com#",
  "distance": 1,
  "referrer": "https://www.bbc.com",
  "referrer-element": "<a id=\"notification-link\" class=\"js-notification-link animated three\" href=\"#\" style=\"display: none\"><span class=\"hidden-span\">Notifications</span><div class=\"notification-link--triangle\"/><div class=\"notification-link--triangle\"/><span id=\"not-num\"/></a>",                                                                                     
  "status": 200,
  "request-time": 169977,
  "exception": null
}
{
  "url": "https://www.bbc.co.uk",
  "distance": 1,
  "referrer": "https://www.bbc.com",
  "referrer-element": "<a href=\"https://www.bbc.co.uk\">Homepage</a>",
  "status": 200,
  "request-time": 2542733,
  "exception": null
}
{
  "url": "http://www.iana.org/domains/example",
  "distance": 1,
  "referrer": "http://www.example.com",
  "referrer-element": "<a href=\"http://www.iana.org/domains/example\">More information...</a>",                                                                                              
  "status": 200,
  "request-time": 6400768,
  "exception": null
}
```

### CODE REVIEW

**Comment by dantleech** [comment]
cc @gitressa

**Comment by gitressa** [comment]
This looks great, and will be really useful in the e-mailed link reports, helping find the broken link faster when using fx CKEditor. I am looking forward to experimenting with it, once it has been merged into Master. Thanks @dantleech!

**Comment by gitressa** [comment]
I just checked out the PR and it works well, great work! There are a few issues:

**Too much data**
The use case in my original request from #49 was intended to improve scannability of the report, and helping to find the offending link, when editing a page with fx CKEditor.

I think this PR includes too much data and actually reduces scannability, since I don't need the mark up of the link, but only this:
- The referring page
- The broken link
- The text of the broken link

**Current patch:**
```
{
  "url": "https://www.bbc.com#",
  "distance": 1,
  "referrer": "https://www.bbc.com",
  "referrer-element": "<a id=\"notification-link\" class=\"js-notification-link animated three\" href=\"#\" style=\"display: none\"><span class=\"hidden-span\">Notifications</span><div class=\"notification-link--triangle\"/><div class=\"notification-link--triangle\"/><span id=\"not-num\"/></a>",                                                                                     
  "status": 200,
  "request-time": 169977,
  "exception": null
}
```
**Better:**
```
{
  "url": "https://www.bbc.com#",
  "distance": 1,
  "referrer": "https://www.bbc.com",
  "referrer-element": "Notifications",                                                                                     
  "status": 200,
  "request-time": 169977,
  "exception": null
}
```
**File size**
The output file size almost doubles and goes up from 4 to 7 Kb, compared to without the referrer-element. Would it be possible to make this field optional in the output-file, if the user doesn't need it? This would be solved if only the link text, but not the mark up was included.

**jq issue with dashes in referrer-element**
I am getting this message if I try to create a report with jq:
```
$ cat linkreport.json | jq -c '. | select(.status==200) | {referrer-element: .referrer-element, url: .url, status: .status}'
jq: error: syntax error, unexpected '-', expecting '}' (Unix shell quoting issues?) at <top-level>, line 1:
. | select(.status==200) | {referrer-element: .referrer-element, url: .url, status: .status}                                    
jq: 1 compile error
```
It seems like jq doesn't like dashes, but if I rename `referrer-element` to `referrer_element`, I can pull the data, so perhaps rename it to that?

**Comment by dantleech** [comment]
> Too much data

Yeah this is true - the idea was that by showing the HTML element it would identify the exact link (e.g. you might have 50 `Read more ...` links on a page, but maybe we could solve this in the future in a better way (e.g. allow the user to select a format).

For now I'd also be happy to remove the HTML and just show the text.

> jq issue with dashes in referrer-element

I'm not sure what the exact issue is, but if it works with `_` then we should use that instead.

**Comment by gitressa** [comment]
I am glad to hear that link text without HTML is possible. And yes, `Read more ...` and image links might need some more work later on to be useful ...

About dash in referrer-element, perhaps this explains it? From the j[q  Manual](https://stedolan.github.io/jq/manual/#Basicfilters):
> The simplest useful filter is .foo. 
...
This syntax only works for simple, identifier-like keys, that is, keys that are all made of alphanumeric characters and underscore, and which do not start with a digit.

**Comment by dantleech** [comment]
How about this:

**Comment by dantleech** [comment]
How about this:

```
{
  "url": "https://github.com/Microsoft/language-server-protocol/blob/gh-pages/specification.md",
  "distance": 2,
  "referrer": "https://www.dantleech.com/page/2",
  "referrer_element": {
    "title": "Langauge Server\nProtocol",
    "xpath": "/html/body/main/div/div/article[1]/div/p[29]/a[2]"
  },
  "status": 200,
  "request_time": 723263,
  "exception": null
}
{
  "url": "https://www.amazon.com/Implementing-Domain-Driven-Design-Vaughn-Vernon/dp/0321834577",
  "distance": 2,
  "referrer": "https://www.dantleech.com/page/2",
  "referrer_element": {
    "title": "Implementing Domain-Driven\nDesign",
    "xpath": "/html/body/main/div/div/article[1]/div/p[17]/a"
  },
  "status": 200,
  "request_time": 1851147,
  "exception": null
}
```

File: lib/Model/Crawler.php [hunk 1]
```diff
@@ -43,14 +45,15 @@ public function crawl(Url $documentUrl, UrlQueue $queue, ReportBuilder $report):
...
+            assert($linkElement instanceof DOMElement);
             $href = $linkElement->getAttribute('href');
 
             if (!$href) {
                 continue;
             }
 
             try {
-                $url = $documentUrl->resolveUrl($href);
+                $url = $documentUrl->resolveUrl($href, ReferringElement::fromDOMElement($linkElement));
```
Comment (dantleech) [inline lib/Model/Crawler.php]: we don't need to pass the link separately here, as it's part of the "referring element". This class needs refactoring.

**Comment by gitressa** [comment]
I have just tried the new code, and I like the new format, it's exactly what I need.

**Comment by gitressa** [comment]
When this PR is merged, quite a few improvements have been added, so would a fresh fink.phar release make sense?

**Comment by dantleech** [comment]
I often get empty links on our website:
```
"referrer_title": "\n                    \n        \n                "
```

**Comment by gitressa** [comment]
It sounds like the link-as-text feature needs some more tweaking ...

Would a fresh release be possible with the current Master? I can always expand the report, when the link text is ready, but the current Master branch of Fink is ready for production, as far as I am concerned, with all the other improvements.

**Comment by dantleech** [comment]
yep, am planning to release any day now when I get a minute ...

**Comment by gitressa** [comment]
Cool, thanks!

**Comment by gitressa** [comment]
This is a great feature, thanks for adding it. I have tested it, and it works fine. `referrer_title` is empty if the link is an image, but that's expected, and shouldn't be a blocker for next release, in my opinion.

**Pull request merged**